web3/account.py

```diff
@@ -0,0 +1,194 @@
+        message_bytes = to_bytes(data, hexstr=hexstr, text=text)
+        recovery_hasher = compose(to_hex, keccak, signature_wrapper)
+        return recovery_hasher(message_bytes)
+
+    def privateKeyToAccount(self, private_key):
+        key_bytes = hexstr_if_str(to_bytes, private_key)
+        try:
+            key_obj = self._keys.PrivateKey(key_bytes)
+            return LocalAccount(key_obj, self.web3)
+        except ValidationError as original_exception:
+            raise_from(
+                ValueError(
+                    "The private key must be exactly 32 bytes long, instead of "
+                    "%d bytes." % len(key_bytes)
+                ),
+                original_exception
+            )
+
+    def recover(self, msghash, vrs=None, signature=None):
+        hash_bytes = hexstr_if_str(to_bytes, msghash)
+        if vrs is not None:
+            v, r, s = map(functools.partial(hexstr_if_str, to_decimal), vrs)
+            v_standard = to_standard_v(v)
+            signature_obj = self._keys.Signature(vrs=(v_standard, r, s))
+        elif signature is not None:
+            signature_bytes = hexstr_if_str(to_bytes, signature)
+            signature_obj = self._keys.Signature(signature_bytes=signature_bytes)
+        else:
+            raise TypeError("You must supply the vrs tuple or the signature bytes")
+        pubkey = signature_obj.recover_public_key_from_msg_hash(hash_bytes)
+        return pubkey.to_checksum_address()
+
+    def recoverMessage(self, data=None, hexstr=None, text=None, vrs=None, signature=None):
+        msg_hash = self.hashMessage(data, hexstr=hexstr, text=text)
+        return self.recover(msg_hash, vrs=vrs, signature=signature)
+
+    def recoverTransaction(self, serialized_transaction):
+        txn_bytes = hexstr_if_str(to_bytes, serialized_transaction)
+        txn = Transaction.from_bytes(txn_bytes)
+        chain_aware_txn = annotate_transaction_with_chain_id(txn)
+        msg_hash = chain_aware_txn.hash()
+        if sys.version_info.major < 3:
+            msg_hash = to_hex(msg_hash)
+        return self.recover(msg_hash, vrs=vrs_from(txn))
+
+    def setKeyBackend(self, backend):
+        self._keys = KeyAPI(backend)
+
+    def sign(self, message=None, private_key=None, message_hexstr=None, message_text=None):
+        '''
+        @param private_key in bytes, str, or int.
+            In Python 2, a bytes, unicode or str object will be interpreted as hexstr
+            In Python 3, only a str object will be interpreted as hexstr
+        '''
+        msg_bytes = to_bytes(message, hexstr=message_hexstr, text=message_text)
+        msg_hash = self.hashMessage(msg_bytes)
+        key_bytes = hexstr_if_str(to_bytes, private_key)
+        key = self._keys.PrivateKey(key_bytes)
+        (v, r, s, eth_signature_bytes) = sign_message_hash(key, msg_hash)
+        (r_hex, s_hex, eth_signature_hex) = map(to_hex, (r, s, eth_signature_bytes))
+        return AttributeDict({
```

Comment:
This makes me think about whether the accounts API should be done using a middleware that listens for non-standard JSON-RPC calls (`accounts_sign`, `accounts_recover`).

Do you have thoughts on this?  Is it weird to you that all of the other API's can be mucked with by middlewares, but this one stands separate?

Reply:
The attrdict middleware is a cute way to handle this line, but I don't think other middlewares buy us anything (`pythonic` -- neutral, `stalecheck` -- negative, others ??). It also introduces a risk if the user removes the default middlewares: in some cases, I think web3 would send their private key to the provider(s) over http. *yikes*

`web3.eth.account` does stand out as different, since it doesn't call the client. I think it **should** stand out as different. We want to be absolutely clear with users that this module won't send your private key anywhere.

In fact, I'm starting to think we should move the default prefills out of `Account.signTransaction()`. That is the only method that is currently making provider calls. Then we can simply say the `eth.account` module has absolutely no calls to the provider.

*Downside:*

 * It becomes slightly less convenient, when signing a transaction without all the details
 * It becomes slightly less API-equivalent with web3.js
   * We would invent some new function like `w3.eth.fillTransactionDefaults({....})`
   * `Account.signTransaction()` would no longer accept a transaction missing any fields

*Upside:*

 * It's much clearer to say "`web3.eth.account` never makes client calls"
 * We could even strip out the `web3` attribute from `Account` to be as clear as possible, and avoid future changes that accidentally re-introduce client calls

Thoughts?

Reply:
> I think web3 would send their private key to the provider(s) over http. yikes

This line sold me. I'm now firmly in the camp that this shouldn't be mock RPC'd

Reply:
> In fact, I'm starting to think we should move the default prefills out of `Account.signTransaction()`

👍

Reply:
> `Account.signTransaction()` would no longer accept a transaction missing any fields

I think that rather than a global `w3.eth.fillTransactionDefaults({....})` we should instead implement localized APIs on `web3.eth`, `web3.eth.accounts`, The `account` object that comes back from `web3.eth.accounts`, and `contract`.  These would all be the same API.  Something like

`web3.eth.setTransactionDefaults({...})`.  This would replace `web3.eth.defaultAccount` (or that could just be a shortcut for `web3.eth.setTransactionDefaults({'from': '0x1234...'})`.

So *everything* that sends transactions or creates transactions get's the same API and we avoid a global set of transaction defaults because I don't actually think that makes sense.

Reply:
> we should instead implement localized APIs on `web3.eth`, `web3.eth.accounts`, The `account` object that comes back from `web3.eth.accounts`, and `contract`.

I don't fully understand the proposal. You're saying we'd have a `web3.eth.accounts.create().setTransactionDefaults()` too?

One thing that might be obscuring the issue is that many of the defaults aren't static, but dynamic from the client. To highlight the "defaults":
```
TRANSACTION_DEFAULTS = {
    'gasPrice': lambda web3: web3.eth.gasPrice,
    'value': lambda web3: 0,
    'data': lambda web3: b'',
    'chainId': lambda web3: int(web3.net.version),
}
# implemented elsewhere:
nonce = web3.eth.getTransactionCount(account.address)
```

So having those get set anywhere in the `web3.eth.account` module (including the `LocalAccount` object returned by a `web3.eth.account.create()`) would involve a client call, which I'd love to prohibit. Although, `value` and `data` could still get defaulted in `web3.eth.account`.

I *think* the rest are globally default-able. `nonce` is the biggest question mark.